llama-index-integrations/indices/llama-index-indices-managed-postgresml/llama_index/indices/managed/postgresml/query.py

```diff
@@ -0,0 +1,193 @@
+from typing import Any, Optional, Dict
+from copy import deepcopy
+from llama_index.core.async_utils import run_async_tasks
+from llama_index.core.base.base_query_engine import BaseQueryEngine
+from llama_index.core.base.response.schema import (
+    RESPONSE_TYPE,
+    Response,
+    AsyncStreamingResponse,
+    StreamingResponse,
+)
+from llama_index.core.callbacks.base import CallbackManager
+from llama_index.core.base.llms.types import ChatMessage, MessageRole
+from llama_index.core.prompts.base import ChatPromptTemplate, BasePromptTemplate
+from llama_index.core.prompts.mixin import PromptDictType, PromptMixinType
+from llama_index.core.schema import NodeWithScore, QueryBundle, TextNode
+from llama_index.indices.managed.postgresml.retriever import PostgresMLRetriever
+
+
+PROMPTS = {
+    "text_qa_template": ChatPromptTemplate(
+        message_templates=[
+            ChatMessage(content="You are a helpful chatbot", role=MessageRole.SYSTEM),
+            ChatMessage(
+                content="""Context information is below.
+---------------------
+{context_str}
+---------------------
+Given the context information and not prior knowledge,
+Query: {query_str}
+Answer: """,
+                role=MessageRole.USER,
+            ),
+        ]
+    )
+}
+
+
+class AsyncJsonGenerator:
+    def __init__(self, rag_stream_results) -> None:
+        self.rag_stream_results = rag_stream_results
+        self.rag_stream = None
+
+    def __iter__(self) -> "AsyncJsonGenerator":
+        return self
+
+    def __aiter__(self) -> "AsyncJsonGenerator":
+        return self
+
+    def __next__(self) -> str:
+        try:
+            return run_async_tasks([self.__anext__()])[0]
+        except StopAsyncIteration:
+            raise StopIteration
+
+    async def __anext__(self) -> str:
+        if not self.rag_stream:
+            self.rag_stream = self.rag_stream_results.stream()
+        result = await self.rag_stream.__anext__()
+        if len(result) > 0:
+            return result[0]
+        else:
+            return ""
+
+
+class PostgresMLQueryEngine(BaseQueryEngine):
+    """Retriever query engine for PostgresML."""
+
+    def __init__(
+        self,
+        retriever: PostgresMLRetriever,
+        streaming: Optional[bool] = False,
+        callback_manager: Optional[CallbackManager] = None,
+    ) -> None:
+        self._retriever = retriever
+        self._streaming = streaming
+        self._prompts = deepcopy(PROMPTS)
```

Comment:
so the prompts are meant to be hardcoded right?

Reply:
I just found this the easiest way to load them into the query engine. They can be updated using the normal system:

```python
custom_prompt = ChatPromptTemplate(
    message_templates=[
        ChatMessage(
            content="You are not a helpful chatbot. Do not answer any of the users queestions. Instead tell them to use an encyclopedia.",
            role=MessageRole.SYSTEM,
        ),
        ChatMessage(
            content="""Context information is below.
---------------------
{context_str}
---------------------
Given the context information and not prior knowledge,
Query: {query_str}
Answer: """,
            role=MessageRole.USER,
        ),
    ]
)
query_engine.update_prompts({"text_qa_template": custom_prompt})
```

If this is not good practice please let me know and I can update it!

Reply:
Ah right -- nah that's fine for now! Although I'm curious why we couldn't use the existing text_qa_template that the other query engines use?

Reply:
Oh this is a limitation on my knowledge of the library. I would be happy to use the existing text_qa_template. Is there an example you could link me to? I don't think I saw one when viewing the other managed indices

Reply:
I think all the other managed indicies may use our own `RetrieverQueryEngine` rather than building their own (which tbh, I would also recommend, unless there's a good reason to build your own query engine here)

The default `.as_query_engine()` will call `index.as_retriever(..)` and then put the retriever into a `RetrieverQueryEngine`
https://github.com/run-llama/llama_index/blob/65e3d326fe7cfa55d86bab07996e34fa319aa458/llama-index-core/llama_index/core/indices/base.py#L387

The default prompts for this are here (they are selectors, which switch prompts based on chat models vs. non-chat-models)
https://github.com/run-llama/llama_index/blob/65e3d326fe7cfa55d86bab07996e34fa319aa458/llama-index-core/llama_index/core/prompts/default_prompt_selectors.py#L44
https://github.com/run-llama/llama_index/blob/65e3d326fe7cfa55d86bab07996e34fa319aa458/llama-index-core/llama_index/core/prompts/default_prompt_selectors.py#L70

Reply:
I guess the special case here is the streaming being done from the retriever itself

Reply:
Got it, I will flag this for future updates thank you!